### ISSUE TITLE
Since loss is pixel wise and vessels are small, weigh vessel classification more strongly in loss function during training

### ISSUE THREAD

**Issue by abuzarmahmood** [pulakatlab/auto_slide#49]

**Comment by abuzarmahmood** [comment]
### Summary of Suggestions and Changes

1. **Files Identified for Modification:**
   - **`training_utils.py`**: Contains utility functions related to training, including the loss function configuration.
   - **`training.py`**: Responsible for the main training process, including the training loop.
   - **`test_training.py`**: Contains tests related to the training process.

2. **Plan to Address the Issue:**
   - Modify the loss function to weigh vessel classification more strongly.
   - Ensure the training loop uses the modified loss function.
   - Update tests to reflect changes in the loss function.

3. **Specific Changes Suggested:**

   1. **Modify Loss Function in Training**
      - **File**: `training_utils.py`
      - **Line**: Around 161-165
      - **Description**: Adjust the loss calculation to weigh vessel classification more strongly by multiplying the vessel classification loss by a factor greater than 1.
      - **Code Snippet**:
        ```python
        # Assuming 'loss' is a dictionary containing different loss components
        vessel_weight = 2.0  # Example weight for vessel classification
        loss['loss_classifier'] *= vessel_weight
        losses = sum([l for l in loss.values()])
        ```

   2. **Ensure Training Loop Uses Modified Loss**
      - **File**: `training.py`
      - **Line**: Ensure that the training loop (lines 95-108) uses the modified loss function.
      - **Description**: No specific code change needed if the loss function is modified in `training_utils.py`. Just ensure that the training loop calls the modified loss function.

   3. **Update Tests to Reflect Changes**
      - **File**: `test_training.py`
      - **Line**: Around 214-221
      - **Description**: Update or add tests to ensure that the modified loss function behaves as expected.
      - **Code Snippet**:
        ```python
        def test_modified_loss_function():
            model = initialize_model()
            model.train()
            img = torch.rand(3, 100, 100)
            target = {
                'boxes': torch.tensor([[10, 10, 50, 50]], dtype=torch.float32),
                'labels': torch.tensor([1], dtype=torch.int64),
                'masks': torch.zeros(1, 100, 100, dtype=torch.uint8)
            }
            target['masks'][0, 20:40, 20:40] = 1
            loss_dict = model([img], [target])
            assert 'loss_classifier' in loss_dict
            assert loss_dict['loss_classifier'] > 0  # Ensure it's being calculated
        ```

These changes aim to ensure that vessel classification is given more importance in the loss function during training. Testing these changes thoroughly is recommended to ensure they have the desired effect.

---
*This response was automatically generated by blech_bot using model gpt-4o*

**Comment by abuzarmahmood** [comment]
Evidence that may change the priority here, tracked under #116.

The premise — that small vessels are under-detected — is confirmed by the
current evaluation: 77% of images in the smallest ground-truth-area quintile
score zero, against 2% in the largest.

But loss weighting may not be the right lever, because a large part of that gap
appears to be upstream of the loss entirely. `AugmentedCustDat.__getitem__`
discards any instance covering less than 0.5% of the image before constructing
the target:

```python
if np.mean(mask == obj) > 0.005:
    fin_objs.append(obj)
```

Measured across all 407 masks: **173 of 646 annotated instances (26.8%) never
reach the model**, and 23 vessel-containing images lose every label and become
negative samples. Evaluation applies no such filter, so the model is trained on
73% of the labels and scored against 100%.

In other words the network is not failing to learn small vessels so much as
being trained not to predict them. Filed as #117.

Suggestion: land #117 and re-measure the size-stratified accuracy before
investing in a weighted loss. If a size bias survives the label fix, this issue
becomes the right next step and the measurement will show how much is left to
recover.
